File: clients/java/src/main/java/net/weweave/tubewarder/client/TubewarderClient.java
package net.weweave.tubewarder.client;

/**
 * An abstract class for sending messages to a Tubewarder server.
 * The concrete implementations define the used protocol and libs.
 */
public abstract class TubewarderClient {
    private final String uri;

    /**
     * @param uri The base uri to the Tubewarder server (i.e. https://localhost/)
     */
    public TubewarderClient(String uri) {
        this.uri = (uri.endsWith("/") ? uri : uri + "/");
    }

    /**
     * Sends a concrete message
     * @param request
     * @return
     */
    public abstract SendResponse send(SendRequest request);

    public String getUri() {
        return uri;
    }
}

File: clients/php/TubewarderClient.php
<?php
namespace Tubewarder;
require('SendRequest.php');
require('SendResponse.php');
require('ErrorCode.php');
require('KeyValue.php');

abstract class TubewarderClient {
    private $uri;

    public function __construct($uri) {
        $this->uri = $uri;
        if (substr($this->uri, strlen($this->uri)-1) !== '/') {
            $this->uri .= '/';
        }
    }

    /**
     * @return string
     */
    public function getUri() {
        return $this->uri;
    }

    /**
     * @param SendRequest $request
     * @return SendResponse
     */
    public abstract function send(SendRequest $request);
}
?>